### PR TITLE
feat(blackfire): Add bridge for upscale/swoole-blackfire

### PR DESCRIPTION
**Pull request by luca-nardelli**
Hi again! 

I noticed that I was unable to profile requests with blackfire (before swoole I was using the standard nginx + php-fpm setup), as `blackfire curl` only told me that the probe was not initialized. After a while I stumbled upon this library https://github.com/upscalesoftware/swoole-blackfire and I tried to connect it to this bundle.

I've managed to make it work (I had to add the WithProfiler configurator after WithRequestHandler otherwise the instrumentation call would fail), however I didn't write any tests at the moment so I'd say the PR cannot be merged yet. I'm opening this PR to discuss this feature (with the proposed implementation)

Question 1: Is this a relevant addition to the bundle?
Question 2: Could you tell me which test I should look at if I wanted to write one for this new feature?

Thanks!

### CODE REVIEW

File: src/Bridge/Upscale/Blackfire/WithProfiler.php [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+use Upscale\Swoole\Blackfire\Profiler;
+
+final class WithProfiler implements ConfiguratorInterface
+{
+    /**
+     * @var Profiler
+     */
+    private $profiler;
+
+    public function __construct()
```
Comment (k911) [inline src/Bridge/Upscale/Blackfire/WithProfiler.php:18]:
```suggestion
    public function __construct(Profiler $profiler)
```

Then, you could write unit tests with the help of prohpecy

Comment (luca-nardelli) [reply src/Bridge/Upscale/Blackfire/WithProfiler.php:18]: Thanks, I'll look into it

**Comment by k911** [comment]
This feature seems to be useful, so yeah feel free to contribute :)

File: composer.json [hunk 2]
```diff
@@ -45,13 +45,14 @@
...
         "symfony/messenger": "^4.3.1|^5.0",
+        "symfony/monolog-bridge": "^4.3.1|^5.0",
+        "symfony/monolog-bundle": "^3.3",
         "symfony/twig-bundle": "^4.3.1|^5.0",
         "symfony/var-dumper": "^4.3.1|^5.0",
-        "symfony/yaml": "^4.3.1|^5.0"
+        "symfony/yaml": "^4.3.1|^5.0",
+        "upscale/swoole-blackfire": "^3.0"
     },
     "suggest": {
```
Comment (k911) [inline composer.json:57]:
```suggestion
    "suggest": {
      "upscale/swoole-blackfire": "To use blackfire" # or something like that,
```

**Comment by luca-nardelli** [comment]
I've added the following:
- A documentation entry in the docs folder
- A Unit test with prophpecy that ensures instrument() is called
- A Feature test that ensures that the WithProfiler configurator is added after WithRequestHandler in the container. To do that, I had to make two services public for testing purposes though.
- An entry in the suggests section of composer.json

Let me know if something else should be changed!

File: tests/Feature/SwooleProfilerRegisteredTest.php [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+use K911\Swoole\Server\Configurator\WithRequestHandler;
+use K911\Swoole\Tests\Fixtures\Symfony\TestBundle\Test\ServerTestCase;
+use ReflectionClass;
+
+final class SwooleProfilerRegisteredTest extends ServerTestCase
+{
+    /**
+     * Ensure that WithProfiler is added after WithRequestHandler.
+     */
+    public function testWiring(): void
```
Comment (k911) [inline tests/Feature/SwooleProfilerRegisteredTest.php:18]: I do not think we need to test this deeply. We should only check whether blackfire profiler and configurator has been registered in container. The ordering of configurations could be a bug, that needs to be fixed in separate PR, and will probably disappear after the refactor (different approach to configure events and swoole server - branch `feat/dynamic-listeners`).

File: tests/Fixtures/Symfony/TestAppKernel.php [hunk 4]
```diff
@@ -109,6 +110,21 @@ protected function configureContainer(ContainerBuilder $c, LoaderInterface $load
         if ($this->coverageEnabled && 'cov' !== $this->environment) {
             $loader->load($confDir.'/cov/**/*'.self::CONFIG_EXTENSIONS, 'glob');
         }
+
+        $c->addCompilerPass(
```
Comment (k911) [inline tests/Fixtures/Symfony/TestAppKernel.php:114]: Correct, me if I'm wrong but there is an available special container that allows to access private services in test suites, so we do not need that.

https://symfony.com/blog/new-in-symfony-4-1-simpler-service-testing
https://symfony.com/doc/current/testing.html#accessing-the-container

File: tests/Unit/Server/Configurator/WithProfilerTest.php [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+/**
+ * @runTestsInSeparateProcesses
+ */
+class WithProfilerTest extends TestCase
+{
+    use \Prophecy\PhpUnit\ProphecyTrait;
+    /**
+     * @var NoOpTaskFinishedHandler
+     */
+    private $noOpTaskFinishedHandler;
```
Comment (k911) [inline tests/Unit/Server/Configurator/WithProfilerTest.php:23]: This property is unused right?

**Comment by k911** [comment]
@luca-nardelli Thanks, really good job! I've left some comments, we'll have to address them and I think that's it.

File: tests/Unit/Server/Configurator/WithProfilerTest.php (hunk 5 above)
Comment (luca-nardelli) [reply tests/Unit/Server/Configurator/WithProfilerTest.php:23]: yup! removed, thanks

File: tests/Fixtures/Symfony/TestAppKernel.php (hunk 4 above)
Comment (luca-nardelli) [reply tests/Fixtures/Symfony/TestAppKernel.php:114]: There is, somehow I never noticed the green box mentioning the `test.service_container` id, thanks for pointing this out!

File: tests/Feature/SwooleProfilerRegisteredTest.php (hunk 3 above)
Comment (luca-nardelli) [reply tests/Feature/SwooleProfilerRegisteredTest.php:18]: I don't understand this, what do you mean that the ordering could be a bug? The WithRequestHandler is added manually to the list of configurators, not via autoconfiguration, so the ordering seems to be fixed to me (if I autoconfigure WithProfiler instead of adding it manually, it's always triggered earlier).

I had to ensure that the profiler was enabled after registering the request handler because otherwise blackfire would throw an exception related to missing middleware or something (can't remember at the moment sorry), that's why I was testing for this, but I can remove it if you want

**Comment by luca-nardelli** [comment]
Fixed the usage of the service container (thanks for that!) and removed the unused property!

File: tests/Fixtures/Symfony/TestAppKernel.php [hunk 6]
```diff
@@ -14,6 +14,7 @@
 use Symfony\Bundle\MonologBundle\MonologBundle;
 use Symfony\Bundle\TwigBundle\TwigBundle;
 use Symfony\Component\Config\Loader\LoaderInterface;
+use Symfony\Component\DependencyInjection\Compiler\CompilerPassInterface;
```
Comment (k911) [inline tests/Fixtures/Symfony/TestAppKernel.php:17]: Some leftovers, I think

File: tests/Feature/SwooleProfilerRegisteredTest.php (hunk 3 above)
Comment (k911) [reply tests/Feature/SwooleProfilerRegisteredTest.php:19]: > I had to ensure that the profiler was enabled after registering the request handler because otherwise blackfire would throw an exception related to missing middleware or something (can't remember at the moment sorry), that's why I was testing for this, but I can remove it if you want

This part you're describing is a probably buggy behaviour in swoole-bundle itself, so instead of making sure the ordering of configurators is right in this test, we should focus on fixing that issue or more like understanding it first - but it is not the purpose of this PR.

I'll comment on lines which I think are unnecessary

File: tests/Feature/SwooleProfilerRegisteredTest.php [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+                    && ($configurator instanceof WithProfiler || $configurator instanceof WithRequestHandler)
+                ) {
+                    $configuratorClasses[] = \get_class($configurator);
+                }
+            }
+            $this->assertEquals([
+                WithRequestHandler::class,
+                WithProfiler::class,
+            ], $configuratorClasses);
+        };
```
Comment (k911) [inline tests/Feature/SwooleProfilerRegisteredTest.php:44]:
```suggestion
```

File: tests/Feature/SwooleProfilerRegisteredTest.php [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+
+        $this->assertInstanceOf(Profiler::class, $testContainer->get(Profiler::class));
```
Comment (k911) [inline tests/Feature/SwooleProfilerRegisteredTest.php:46]:
```suggestion
        $this->assertTrue($testContainer->has(Profiler::class));
```

Should be enough

File: tests/Feature/SwooleProfilerRegisteredTest.php [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+
+        $inspectionFn($testContainer->get('swoole_bundle.server.http_server.configurator.for_server_run_command'));
+        $inspectionFn($testContainer->get('swoole_bundle.server.http_server.configurator.for_server_start_command'));
```
Comment (k911) [inline tests/Feature/SwooleProfilerRegisteredTest.php:49]:
```suggestion
       $this->assertTrue($testContainer->has(WithProfiler::class));
```

File: tests/Feature/SwooleProfilerRegisteredTest.php (hunk 3 above)
Comment (luca-nardelli) [reply tests/Feature/SwooleProfilerRegisteredTest.php:19]: Ok now I understood, thanks!

File: tests/Fixtures/Symfony/TestAppKernel.php (hunk 6 above)
Comment (luca-nardelli) [reply tests/Fixtures/Symfony/TestAppKernel.php:17]: Yup, thanks

**Comment by luca-nardelli** [comment]
Ok I should have applied all your suggestions

**Comment by k911** [comment]
@luca-nardelli Thanks again for your hard work, I'll leave this on `develop` branch for a bit, then release.

**Comment by luca-nardelli** [comment]
Great! Thank you for the bundle :)

Il sab 23 mag 2020, 14:45 Konrad Obal <notifications@github.com> ha scritto:
